File: Data Structures & Algorithms/three-integer-sum/submission-0.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        ans = []
        nums.sort()

        for i in range(len(nums) - 2):
            # Skip duplicate values for i
            if i > 0 and nums[i] == nums[i - 1]:
                continue

            l, r = i + 1, len(nums) - 1
            while l < r:
                s = nums[i] + nums[l] + nums[r]
                if s == 0:
                    ans.append([nums[i], nums[l], nums[r]])

                    # Skip duplicates for l
                    while l < r and nums[l] == nums[l + 1]:
                        l += 1
                    # Skip duplicates for r
                    while l < r and nums[r] == nums[r - 1]:
                        r -= 1

                    l += 1
                    r -= 1

                elif s < 0:
                    l += 1
                else:
                    r -= 1

        return ans

File: Data Structures & Algorithms/combination-target-sum-ii/submission-0.cpp
class Solution {
public:
    vector<vector<int>> res;

    void solve(int indx, vector<int> &path, vector<int> &nums, int rem) {
        if (rem == 0) {
            res.push_back(path);
            return;
        }

        for (int i = indx; i < nums.size(); i++) {
            // skip duplicates
            if (i > indx && nums[i] == nums[i - 1]) continue;

            if (nums[i] > rem) break;

            path.push_back(nums[i]);
            solve(i + 1, path, nums, rem - nums[i]);
            path.pop_back();
        }
    }

    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<int> path;
        solve(0, path, candidates, target);
        return res;
    }
};